### ISSUE TITLE
cannot change type of sumtype in an inline match statement

### ISSUE THREAD

**Issue by Hunam6** [vlang/v#13774]
```
OS: macos, macOS, 12.3, 21E230
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.0.0 (clang-1300.0.29.30)

getwd: /Users/hunam/Documents/dev/go2v
vmodules: /Users/hunam/.vmodules
vroot: /Applications/v
vexe: /Applications/v/v
vexe mtime: 2022-03-19 12:45:57
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 f7feb63.0cc6310

Git version: git version 2.32.0 (Apple Git-132)
Git vroot status: weekly.2022.11-28-g0cc63107
.git/config present: true
thirdparty/tcc status: thirdparty-unknown-unknown de82a130
```

**What did you do?**
```v
struct Aa {}

struct Bb {}

type Sumtype = Aa | Bb

fn main() {
	mut var := Sumtype(Aa{})

	if mut var is Aa {
		// works
		var = Bb{}

		// doesn't work
		var = match true {
			true { Bb{} }
			false { Bb{} }
		}
	}
}
```

**What did you expect to see?**
only a warning about the unused variable "var"
 
**What did you see instead?**
```v
ok.v:15:9: error: cannot assign to `var`: expected `Aa`, not `Bb`
   13 |
   14 |         // doesn't work
   15 |         var = match true {
      |               ~~~~~~~~~~~~
   16 |             true { Bb{} }
   17 |             false { Bb{} }
```

**Comment by ntrel** [comment]
Unless this has changed more recently, the error is correct. var is of type Aa within the if block. The outer var is shadowed.
There should however be an error on the var = Bb line.

**Comment by JalonSolov** [comment]
So your claim is that once a variable has been assigned a specific instance of a SumType, the type is frozen to that subtype.

**Comment by ntrel** [comment]
@JalonSolov if you're replying to me, that isn't my claim. My understanding is that the if statement is a smart cast which narrows the type of var. If V has been changed to support 'un'smartcasting in the last year then my reply was wrong. I would like to try it but the playground buttons are disabled and I'm AFK.

**Comment by Hunam6** [comment]
My understanding is that `var := Bb{}` means `var <- Sumtype <- Bb`, it cannot narrow it because otherwise it would mean something like `var <- Bb`.

Furthermore, even if your understanding is "better", what are the drawbacks of allowing what I propose? I personally don't see any.

**Comment by spytheman** [comment]
> Furthermore, even if your understanding is "better", what are the drawbacks of allowing what I propose? I personally don't see any.

Inside the smartcast block (the `if var is Aa {`), V assumes that `var` is of type `Aa`, which is more specific i.e. excluding all the rest of the variants, than the initial `Sumtype`.

The drawback of what you are proposing, is additional complexity: 
V would have to track the original `Sumtype` type of `var`, then allow the assignment, knowing that `var` had originally been of type `Sumtype` and that `Bb` is a variant of `Sumtype`, then do the inferred `var <- Sumtype <- Bb`, and use `Sumtype` for `var`, from that line, till the end of the scope.

TLDR: this is not a bug, but more like a feature request.

**Comment by spytheman** [comment]
Actually, I am wrong - there is a bug there (may be several):
```v
struct Aa {}

struct Bb {}

struct Cc {}

type Sumtype = Aa | Bb | Cc

fn main() {
        mut var := Sumtype(Aa{})
        eprintln('> var: $var | typeof(var): ${typeof(var)} | typeof(var).name: ${typeof(var).name}')
        if mut var is Aa {
                eprintln('> var: $var | typeof(var): ${typeof(var)} | typeof(var).name: ${typeof(var).name}')
                var = Bb{}
                eprintln('> var: $var | typeof(var): ${typeof(var)} | typeof(var).name: ${typeof(var).name}')
                c := match true {
                        true { Cc{} }
                        false { Cc{} }
                }
                var = c
                eprintln('> var: $var | typeof(var): ${typeof(var)} | typeof(var).name: ${typeof(var).name}')
        }
        eprintln('> var: $var | typeof(var): ${typeof(var)} | typeof(var).name: ${typeof(var).name}')
}
```
produces:
```
> var: Sumtype(Aa{}) | typeof(var): Aa | typeof(var).name: Sumtype
> var: Aa{} | typeof(var): Aa | typeof(var).name: Aa
> var: Aa{} | typeof(var): Aa | typeof(var).name: Aa
> var: Aa{} | typeof(var): Aa | typeof(var).name: Aa
> var: Sumtype(Cc{}) | typeof(var): Cc | typeof(var).name: Sumtype
```

**Comment by ntrel** [comment]
> what are the drawbacks of allowing what I propose? 

The typing would be very complex:

```v
	mut var := Sumtype(Aa{})
	if mut var is Aa {
		// here `var` has type Aa
		for i in 0..2 {
			// <-- what's the type of `var` here?
			var = Bb{}
			// here `var` is a Sumtype
		}
```
When i == 0, presumably the type there would still be Aa.
When i == 1, the type would have to be Sumtype.
This is closer to dynamic typing than static typing. It is possible but would make the analysis very complex and slower, plus it would make the language more complex.

**Comment by ntrel** [comment]
I think `var = Bb{}` should be disallowed. Consider the following:
```v
	mut a := Aa{}
	mut p := &a
	mut var := Sumtype(a)
	if mut var is Aa {
		p = &var // currently cgen error
		// works
		var = Bb{}
		p = &var // currently passes checker
```
The cgen error can be fixed (it is just unnessarily dereferencing var which isn't a pointer).
The problem is the second assignment - the checker thinks `var` is still `Aa` when (if allowed) it should have changed to type `Sumtype`.

**Comment by Hunam6** [comment]
I completely understand what you're saying and your point of view.

---

I my own use it was something like this:

I develop [Go2V](https://github.com/vlang/go2v) and a specific step of the processing is to transform "remaining" Go-ish things into V-ish things. eg: `len(array)`->`array.len`.

In my system I have a sumtype containing all the *(current supported)* possible statements (match stmt, if..else stmt, var stmt, etc.).

In this code...
<img width="630" alt="Screenshot 2022-03-21 at 21 45 13" src="https://user-images.githubusercontent.com/38606542/159360981-5948a5fc-6dc5-435a-b22a-6341c1c13309.png">
...the first match branch returns a `BasicValueStmt`, the 2nd a `BasicValueStmt`, the 3rd a `CallStmt`, the 4th a `CallStmt`, the 5th a `CallStmt`, and the last one a `PushStmt`.

I think I could actually assign the result of the `match` statement to a temporary variable and then outside of the `mut stmt is CallStmt` code block I could assign the temporary variable to the `stmt` variable. I guess I could do that but that would not be natural IMHO.

---

All in all I would say that this is a quite complex question that may require more thoughts from other members.

ping @medvednikov @Larpon @JalonSolov, others

**Comment by JalonSolov** [comment]
Personally, I think it should work.  A mutable SumType should allow any of the individual types to be assigned to it _at any time_.  Otherwise, it's not really mutable, is it?

**Comment by ntrel** [comment]
> Personally, I think it should work. A mutable SumType should allow any of the individual types to be assigned to it _at any time_. Otherwise, it's not really mutable, is it?

So presumably you think either:
* my example with `for` should allow var to have different static types on each iteration
* Smart casting should be somehow changed or removed. Currently it changes the static type of var.

**Comment by Larpon** [comment]
I don't think `var = Bb{}` should be allowed in `if mut var is Aa`.
From my understanding `if mut var is Aa` allows for mutating *fields* and call functions that mutate `var`.

I agree with @ntrel that `var = Bb{}` should've given an error.

So:
```v
mut var := Sumtype(Aa{})
if mut var is Aa {
	var.field = true // OK
	var.mutating_receiver() // OK
	var = Bb{} // ERROR
}
```

**Comment by ntrel** [comment]
> if mut var is Aa allows for mutating fields and call functions that mutate var.

Yes, if the Aa instance is not mutated I don't think mut should be required for the smart cast.

**Comment by dumblob** [comment]
>Yes, if the Aa instance is not mutated I don't think `mut` should be required for the smart cast.

IIRC this has been discussed before elsewhere and the consensus was it really shouldn't be required. But nobody has implemented it since then.
